synthesizer/process/src/stack/mod.rs

```diff
@@ -363,10 +351,44 @@ impl<N: Network> StackProgram<N> for Stack<N> {
     /// Returns the expected number of calls for the given function name.
     #[inline]
     fn get_number_of_calls(&self, function_name: &Identifier<N>) -> Result<usize> {
-        self.number_of_calls
-            .get(function_name)
-            .copied()
-            .ok_or_else(|| anyhow!("Function '{function_name}' does not exist"))
+        // Initialize the base number of calls.
+        let mut num_calls = 1;
+        // Initialize a queue of functions to check.
+        let mut queue = vec![(StackRef::Internal(self), *function_name)];
+        // Iterate over the queue.
+        while let Some((stack_ref, function_name)) = queue.pop() {
+            // Ensure that the number of calls does not exceed the maximum.
+            // Note that one transition is reserved for the fee.
+            ensure!(
```

Comment:
I think it can be safely checked once after the while loop.

Reply:
I believe it needs to happen on each loop iteration to be safe.
Otherwise, you may execute for an unnecessarily long time (possibly indefinitely) when traversing large (or cyclic) call graphs.